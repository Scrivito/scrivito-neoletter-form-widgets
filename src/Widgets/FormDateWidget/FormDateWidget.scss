.scrivito-neoletter-form-widgets .form-date {
  .date-title {
    p {
      display: inline;
    }
  }
}
