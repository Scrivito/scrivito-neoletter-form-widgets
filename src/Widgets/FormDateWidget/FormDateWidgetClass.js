import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "string",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"],
      },
    ],
  },
});
