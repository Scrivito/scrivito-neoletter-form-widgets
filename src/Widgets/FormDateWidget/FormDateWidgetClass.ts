import * as Scrivito from "scrivito";

export const FormDateWidget = Scrivito.provideWidgetClass("FormDateWidget", {
  attributes: {
    title: "html",
    helpText: "html",
    required: "boolean",
    customFieldName: "string",
    dateType: [
      "enum",
      {
        values: ["date", "datetime-local"]
      }
    ]
  }
});
