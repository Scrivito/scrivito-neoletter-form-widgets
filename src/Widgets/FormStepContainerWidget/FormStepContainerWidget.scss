// Styles for FormContainerWidget
.scrivito-neoletter-form-widgets.form-container-widget {

  .dropdown-container,
  .select-container {
    .select-title {
      display: flex;
      align-items: baseline;
    }
  }

  .text-super {
    vertical-align: super;
  }

  .text-mandatory {
    font-weight: 700;
    color: var(--bs-danger, #f03a47); //$secondary;
    cursor: default;
  }

  .rating-icon.bi {
    color: grey;
  }

  .winnie-the-pooh {
    opacity: 0;
    position: absolute;
    top: 0;
    left: 0;
    height: 0;
    width: 0;
    z-index: -1;
  }

  .form-buttons {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;

    .step-counter {
      width: 100%;
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: center;
    }

    .review-button {
      margin-right: 1em;
    }
  }

  .hide {
    display: none;
  }

  &.form-border {
    border: black 1px dotted;
    padding: 0.5em;
  }

  .captcha-container {
    display: flex;

    &.left {
      justify-content: left;
    }

    &.center {
      justify-content: center;
    }

    &.right {
      justify-content: right;
    }
  }

  .missing-site-key {
    border: #f03a47 1px solid;
    padding: 1em 0;
  }

  // popover
  .mandatory-container,
  .helptext-container {
    position: relative;
    display: inline-block;

    .form-popover-container {
      position: absolute;
      inset: auto auto 0px 0px;
      transform: translate(-50px, -30px);
      background-color: #fff;
      border: 1px solid #ccc;
      border-radius: 0.5rem;
      box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
      z-index: 1000;
      max-width: 276px;
      width: max-content;

      .form-popover-body {
        font-size: 14px;
        color: #677788;
        padding: 1rem;
        font-style: normal;
        font-weight: 400;
        line-height: 1.5;
        word-wrap: break-word;

        p {
          margin: 0;
        }
      }
    }
  }

  // review modal
  .form-modal {
    position: fixed;
    top: 0;
    left: 0;
    z-index: 1060;
    width: 100%;
    height: 100%;
    overflow-x: hidden;
    overflow-y: auto;
    outline: 0;
    background-color: #212529;

    .modal-dialog {
      width: 80%;
      max-width: 800px;
      margin: 1.75rem auto;

      @keyframes animatetop {
        from {
          top: -300px;
          opacity: 0
        }

        to {
          top: 0;
          opacity: 1
        }
      }

      .modal-content {
        animation: animatetop 0.4s;
        position: relative;
        display: flex;
        flex-direction: column;
        width: 100%;
        pointer-events: auto;
        background-color: #fff;
        background-clip: padding-box;
        border: 1px solid rgba(0, 0, 0, .2);
        border-radius: .3rem;
        outline: 0;

        .modal-header {
          display: flex;
          flex-shrink: 0;
          align-items: center;
          justify-content: center;
          padding: 1rem 1rem;
          border-bottom: 1px solid #dee2e6;
          border-top-left-radius: calc(.3rem - 1px);
          border-top-right-radius: calc(.3rem - 1px);

          h4 {
            margin-bottom: 0;
            line-height: 1.5;
            font-family: "Segoe UI", Arial, sans-serif;
            font-weight: 500;
            margin: 10px 0;
          }
        }

        .modal-body.form-review {
          height: 60vh;
          overflow: auto;
          position: relative;
          flex: 1 1 auto;
          padding: 1rem;
          pointer-events: auto;
          word-break: break-word;

          div.step-review-container {
            &:not(:first-child) {
              .step-review {
                margin-top: 1em !important;
              }
            }
          }

          .review-item-title {
            color: #9e9e9e;
            margin-bottom: 0;
            margin-top: 1em;
            font-size: 1em;
            font-weight: normal;
          }

          .review-item-value {
            //  margin-top: 0.3em;
            font-weight: normal;
            white-space: pre-wrap;
            max-width: 100vw;
            overflow-x: auto;
          }
        }

        .modal-footer {
          display: flex;
          flex-wrap: wrap;
          flex-shrink: 0;
          align-items: center;
          justify-content: flex-end;
          padding: .75rem;
          border-top: 1px solid #dee2e6;
          border-bottom-right-radius: calc(.3rem - 1px);
          border-bottom-left-radius: calc(.3rem - 1px);

          button {
            color: #fff;
            background-color: var(--bs-primary, #5c9dcd);
            border-color: var(--bs-primary, #5c9dcd);
            margin: .25rem;
            padding: .375rem .75rem;
            font-size: 1rem;
            border-radius: .25rem;

            &:hover {
              background-color: var(--bs-btn-hover-bg, #395aa5);
            }
          }
        }
      }
    }
  }

  // Bootstrap Icons - Sizes Add-on
  .bi {
    &.bi-1x {
      font-size: 1em;
      line-height: 1;
    }

    &.bi-default {
      font-size: 1.3em;
      line-height: 1;
    }

    &.bi-2x {
      font-size: 2em;
      line-height: 1;
    }

    &.bi-3x {
      font-size: 3em;
      line-height: 1;
    }

    &.bi-4x {
      font-size: 4em;
      line-height: 1;
    }

    &.bi-5x {
      font-size: 5em;
      line-height: 1;
    }
  }

  @keyframes rotate-icon {
    0% {
      transform: rotate(0deg);
    }

    100% {
      transform: rotate(360deg);
    }
  }

  .rotate-icon {
    display: inline-block;
    animation: rotate-icon 2s linear 0s infinite;
  }
}

.bg-dark-image,
.bg-brand-primary,
.bg-brand-secondary,
.bg-grey,
.bg-greydark,
.bg-greymiddle {

  .missing-id,
  .missing-site-key,
  .scrivito-neoletter-form-widgets.form-container-widget .linear-scale-container,
  .scrivito-neoletter-form-widgets.form-container-widget .reset-label span,
  .scrivito-neoletter-form-widgets.form-container-widget label,
  .scrivito-neoletter-form-widgets.form-container-widget .text-mandatory,
  .scrivito-neoletter-form-widgets.form-container-widget .bi,
  .scrivito-neoletter-form-widgets.form-container-widget .text-super,
  .card.card-theme .scrivito-neoletter-form-widgets.form-container-widget label,
  .card.card-theme .scrivito-neoletter-form-widgets.form-container-widget .text-mandatory,
  .card.card-theme .scrivito-neoletter-form-widgets.form-container-widget .bi {
    color: #fff;
  }

  // hardcoded for now
  .card .scrivito-neoletter-form-widgets.form-container-widget label {
    color: #495054
      /*$theme-grey */
      !important;
  }

  .card .scrivito-neoletter-form-widgets.form-container-widget .text-mandatory {
    color: red !important;
  }

  .card .scrivito-neoletter-form-widgets.form-container-widget .bi {
    color: #ccc !important;
  }
}

.bg-dark-image,
.bg-brand-primary,
.bg-brand-secondary,
.bg-grey,
.bg-greydark,
.bg-greymiddle {
  .scrivito-neoletter-form-widgets.form-container-widget .reset-label:hover {
    background-color: unset;
  }
}

.scrivito-neoletter-form-widgets.missing-id {
  display: flex;
  justify-content: center;
}

// FormIdComponent tab 
.neoletter-form-id-tab-container {
  .scrivito_detail_content {
    border-top: 2px solid rgba(0, 0, 0, 0.1);
    margin: 0;
    padding: 15px;
    font-family: Helvetica Neue, Helvetica, Arial, sans-serif;

    .scrivito_detail_label {
      display: block;
      padding: 0;
      padding-bottom: 4px;
      width: 100%;

      span {
        padding: 7px 3px 6px 3px;
        color: #444;
        display: block;
        font-size: 13px;
        font-weight: bold;
        font-family: "Manrope", sans-serif;
      }

      &.scrivito_error {

        i,
        span {
          color: #c04e4e !important;
        }

        .scrivito_icon_error {
          float: left;
          padding: 9px 3px;
          font-size: 11px;
          font-style: normal;
          font-weight: normal;
          line-height: 13px !important;
          vertical-align: middle;
          text-decoration: none;
          position: relative;
          -webkit-font-smoothing: antialiased;
          -moz-osx-font-smoothing: grayscale;
          -webkit-user-select: none;
          -moz-user-select: none;
          -ms-user-select: none;
          -o-user-select: none;
          user-select: none;
        }
      }
    }

    .content_property_input {
      margin-bottom: 5px;
      box-shadow: 0 0 3px rgba(0, 0, 0, .2);
      border-radius: 3px;
      background: white;
      color: #555;
      display: block;
      font-family: Helvetica Neue, Helvetica, Arial, sans-serif;
      font-size: 14px;
      font-style: normal;
      margin-bottom: 5px;
      padding: 0 8px;
      min-height: 30px;
      line-height: 30px;
      word-break: break-all;

      &.scrivito_error {
        box-shadow: 0 0 0 2px #c04e4e inset;
      }
    }

    .scrivito_validation_notice {
      border-left-width: 7px !important;
      padding: 5px 10px;
      font-size: 11px;
      min-height: 30px;
      line-height: 19px;
      font-weight: normal;
      margin-bottom: 5px;
      box-shadow: 0 0 3px rgba(0, 0, 0, .2);
      border-radius: 3px;
      background: rgba(255, 255, 255, .6);

      &.scrivito_error {
        border-left: 2px solid #c04e4e;
        color: #666;
        box-shadow: 0 0 0 2px #c04e4e inset;
      }
    }

    .scrivito_notice_body {
      color: #666;
      font-weight: bold;
      font-size: 12px;
      padding: 0 0 10px;
    }

    .scrivito_button {
      display: inline-block;
      color: #555;
      font-size: 14px;
      font-weight: normal;
      line-height: 14px;
      text-align: center;
      padding: 8px 10px;
      margin: 0 10px 10px 0;
      position: relative;
      z-index: 1;
      vertical-align: middle;
      min-width: 90px;
      min-height: 14px;
      font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
      background: #fcfcfc;
      border-radius: 3px;
      -webkit-user-select: none;
      -moz-user-select: none;
      -ms-user-select: none;
      -o-user-select: none;
      user-select: none;
      text-decoration: none;

      &.scrivito_disabled {
        color: #aaa !important;
        cursor: not-allowed;
        background: #eee;
      }

      &.scrivito_blue {
        background: #426698 !important;
        color: #fff !important;
      }
    }
  }

  &.scrivito_dark {
    .scrivito_detail_label span {
      color: #eee;
      text-shadow: 0 1px 1px rgba(0, 0, 0, .3), 0 0 4px rgba(0, 0, 0, .2);
    }

    .content_property_input {
      background: rgba(222, 222, 222, .2);
      color: #eee;
    }

    .scrivito_validation_notice {
      color: #eee !important;
      background: rgba(222, 222, 222, 0.2);
    }
  }
}