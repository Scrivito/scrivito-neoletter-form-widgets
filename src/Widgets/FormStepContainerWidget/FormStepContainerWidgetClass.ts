import * as Scrivito from "scrivito";

export const FormStepContainerWidget = Scrivito.provideWidgetClass(
  "FormStepContainerWidget",
  {
    attributes: {
      formId: "string",
      failedMessage: "string",
      submittedMessage: "string",
      submittingMessage: "string",
      hiddenFields: ["widgetlist", { only: "FormHiddenFieldWidget" }],
      formType: [
        "enum",
        {
          values: ["single-step", "multi-step"]
        }
      ],
      steps: ["widgetlist", { only: "FormStepWidget" }],
      forwardButtonText: "string",
      backwardButtonText: "string",
      submitButtonText: "string",
      showBorder: "boolean",
      showReview: "boolean",
      includeEmptyAnswers: "boolean",
      showStepsInReview: "boolean",
      showReviewHeader: "boolean",
      showReviewFooter: "boolean",
      reviewButtonText: "string",
      reviewHeaderTitle: "string",
      reviewCloseButtonText: "string",
      singleSubmitButtonAlignment: [
        "enum",
        {
          values: ["left", "text-center", "text-end", "block"]
        }
      ]
    },
    extractTextAttributes: ["steps"]
  }
);
