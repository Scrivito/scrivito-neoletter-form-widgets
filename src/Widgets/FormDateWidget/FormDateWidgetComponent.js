import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormContainerWidget/utils/getFieldName";
import { HelpText } from "../FormContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormContainerWidget/components/MandatoryComponent";
import "./FormDateWidget.scss";

Scrivito.provideComponent("FormDateWidget", ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")}
        required={widget.get("required")}
      ></input>
      <input type="hidden" name={getFieldName(widget)} value={value}></input>
    </div>
  );
});
