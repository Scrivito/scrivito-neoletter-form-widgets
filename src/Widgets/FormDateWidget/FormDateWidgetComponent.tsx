import * as React from "react";
import * as Scrivito from "scrivito";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { Mandatory } from "../FormStepContainerWidget/components/MandatoryComponent";
import { FormDateWidget } from "./FormDateWidgetClass";
import "./FormDateWidget.scss";

Scrivito.provideComponent(FormDateWidget, ({ widget }) => {
  const [value, setValue] = React.useState("");
  const onChangeValue = (e: React.BaseSyntheticEvent) => {
    const d = new Date(e.target.value);
    setValue(d.toISOString());
  };
  return (
    <div className="form-date mb-3">
      <div className="date-title">
        <span className="text-super">{widget.get("title")}</span>
        {widget.get("required") && <Mandatory />}
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>

      <input
        onChange={onChangeValue}
        className="datepicker"
        type={widget.get("dateType")!}
        required={widget.get("required")}></input>
      <input
        type="hidden"
        className="show-in-review"
        name={getFieldName(widget)}
        value={value}></input>
    </div>
  );
});
