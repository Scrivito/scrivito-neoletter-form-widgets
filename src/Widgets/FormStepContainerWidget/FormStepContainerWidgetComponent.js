import * as React from "react";
import * as Scrivito from "scrivito";
import { scrollIntoView } from "./utils/scrollIntoView";
import { FormFooterMultiSteps } from "./components/FormFooterMultiStepsComponent";
import { FormFooterSingleStep } from "./components/FormFooterSingleStepComponent";
import { FormHiddenFields } from "./components/FormHiddenFieldsComponent";
import { getInstanceId } from "../../config/scrivitoConfig";
import { submitForm } from "./utils/submitForm";
import { FormNoTenant } from "./components/FormNoTenantComponent";
import { FormSubmissionFailed } from "./components/FormSubmissionFailedComponent";
import { FormSubmissionSucceeded } from "./components/FormSubmissionSucceededComponent";
import { FormSubmitting } from "./components/FormSubmittingComponent";
import "./FormStepContainerWidget.scss";

Scrivito.provideComponent("FormStepContainerWidget", ({ widget }) => {
  const tenant = getInstanceId();
  if (!tenant) {
    return <FormNoTenant />;
  }
  const formEndpoint = `https://api.justrelate.com/neoletter/instances/${tenant}/form_submissions`;
  const [currentStep, setCurrentStepNumber] = React.useState(1);
  const [isSubmitting, setIsSubmitting] = React.useState(false);
  const [successfullySent, setSuccessfullySent] = React.useState(false);
  const [submissionFailed, setSubmissionFailed] = React.useState(false);
  const isSingleStep = widget.get("formType") == "single-step";
  const stepsLength = widget.get("steps").length;
  const isLastPage = currentStep == stepsLength;
  const showReview = widget.get("showReview");

  React.useEffect(() => {
    if (!Scrivito.isInPlaceEditingActive()) {
      return;
    }
    // in order to show step number in the props title of each step
    const steps = widget.get("steps");
    steps.forEach((step, i) => {
      const stepNumber = i + 1;
      step.update({
        stepNumber: stepNumber,
        isSingleStep: isSingleStep,
      });
    });
    if (stepsLength > 1 && isSingleStep) {
      widget.update({ formType: "multi-step" });
    } else if (stepsLength == 1 && !isSingleStep) {
      widget.update({ formType: "single-step" });
    }
  }, [widget.get("steps")]);

  if (isSubmitting) {
    return <FormSubmitting submittingText={widget.get("submittingMessage")} />;
  }

  if (successfullySent) {
    return (
      <FormSubmissionSucceeded
        submissionSuccessText={widget.get("submittedMessage")}
      />
    );
  }

  if (submissionFailed) {
    return (
      <FormSubmissionFailed
        submissionFailureText={widget.get("failedMessage")}
      />
    );
  }

  return (
    <div
      className={`scrivito-form-widgets form-container-widget ${
        widget.get("showBorder") ? "form-border" : ""
      }`}
    >
      <form method="post" id={widget.get("formId")}>
        <FormHiddenFields widget={widget} />
        <Scrivito.ContentTag
          content={widget}
          attribute={"steps"}
          widgetProps={{
            getData: (stepId) => {
              const steps = widget.get("steps");
              let isActive = false;
              let stepNumber = 0;
              steps.some((step, index) => {
                if (step.id() == stepId) {
                  stepNumber = index + 1;
                  isActive = stepNumber == currentStep;
                  return true;
                }
              });
              return { stepNumber, isActive, isSingleStep };
            },
          }}
        />
      </form>
      {isSingleStep ? (
        <FormFooterSingleStep widget={widget} onSubmit={onSubmit} />
      ) : (
        <FormFooterMultiSteps
          widget={widget}
          onSubmit={onSubmit}
          onPageChange={onPageChange}
          currentStep={currentStep}
          stepsLength={stepsLength}
          isLastPage={isLastPage}
          showReview={showReview}
        />
      )}
    </div>
  );

  async function onSubmit() {
    if (Scrivito.isInPlaceEditingActive()) {
      return;
    }
    const isValid = validateCurrentStep();
    if (!isValid) {
      return;
    }
    const formElement = document.getElementById(widget.get("formId"));
    scrollIntoView(formElement);

    indicateProgress();
    try {
      await submitForm(formElement, formEndpoint);
      indicateSuccess();
    } catch (e) {
      setTimeout(() => {
        throw e;
      }, 0);

      indicateFailure();
    }
  }

  function indicateProgress() {
    setIsSubmitting(true);
    setSuccessfullySent(false);
    setSubmissionFailed(false);
  }

  function indicateSuccess() {
    setIsSubmitting(false);
    setSuccessfullySent(true);
    setSubmissionFailed(false);
  }

  function indicateFailure() {
    setIsSubmitting(false);
    setSuccessfullySent(false);
    setSubmissionFailed(true);
  }

  function validateCurrentStep() {
    return doValidate(widget.get("formId"), currentStep);
  }

  function onPageChange(next) {
    let isValid = true;

    if (Scrivito.isInPlaceEditingActive()) {
      return;
    }
    if (next) {
      isValid = validateCurrentStep();
    }
    if (!isValid) {
      return;
    }
    const stepNumber = next
      ? Math.min(currentStep + 1, stepsLength)
      : Math.max(currentStep - 1, 1);
    setCurrentStepNumber(stepNumber);
    const formElement = document.getElementById(widget.get("formId"));
    scrollIntoView(formElement);
  }
});

function doValidate(formId, currentStep) {
  let isValid = true;
  const form = document.getElementById(formId);
  if (form) {
    const step = form.querySelector(`[data-step-number='${currentStep}']`);
    if (step) {
      const allInputs = step.querySelectorAll("input, select, textarea") || [];
      for (const node of allInputs.values()) {
        if (!node.checkValidity()) {
          node.reportValidity();
          return (isValid = false);
        }
      }
    }
    return isValid;
  }
}
