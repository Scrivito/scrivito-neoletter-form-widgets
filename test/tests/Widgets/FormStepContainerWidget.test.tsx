import * as Scrivito from "scrivito";
import { screen } from "@testing-library/react";
import { FormStepContainerWidget } from "../../../src/Widgets/FormStepContainerWidget/FormStepContainerWidgetClass";
import { FormStepWidget } from "../../../src/Widgets/FormStepWidget/FormStepWidgetClass";
import { FormDateWidget } from "../../../src/Widgets/FormDateWidget/FormDateWidgetClass";
import "../../../src/Widgets/FormStepContainerWidget/FormStepContainerWidgetComponent";
import "../../../src/Widgets/FormStepWidget/FormStepWidgetComponent";
import "../../../src/Widgets/FormDateWidget/FormDateWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

jest.mock('friendly-challenge', () => ({
  WidgetInstance: jest.fn(() => ({}))
}));
Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();
const widgetProps = {
  formId: "test-id",
  failedMessage: "failed",
  submittedMessage: "submitted",
  submittingMessage: "submitting",
  hiddenFields: [],
  formType: "single-step",
  steps: [],
  forwardButtonText: "forward",
  backwardButtonText: "back",
  submitButtonText: "submit",
  showBorder: false,
  showCaptcha: false,
  captchaAlignment:"center",
  showReview: true,
  includeEmptyAnswers: false,
  showStepsInReview: false,
  showReviewHeader: false,
  showReviewFooter: false,
  reviewButtonText: "Review",
  reviewHeaderTitle: "Review header",
  reviewCloseButtonText: "Close",
  singleSubmitButtonAlignment: "left"
};

describe("FormStepContainerWidget", () => {
  const lodashEsModule = require("lodash-es");

  it("does not render with missing instanceId", () => {
    lodashEsModule.isEmpty.mockReturnValue(true);
    pageRenderer.render({
      body: [new FormStepContainerWidget(widgetProps)]
    });

    const container = document.querySelector("form");
    const emptyTenantElement = document.querySelector(
      ".scrivito-neoletter-form-widgets.missing-id"
    );
    expect(container).not.toBeInTheDocument();
    expect(emptyTenantElement).toBeInTheDocument();
  });

  it("renders with single step mode", () => {
    lodashEsModule.isEmpty.mockReturnValue(false);
    const items = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    const step = new FormStepWidget({ isSingleStep: true, items: items });
    pageRenderer.render({
      body: [new FormStepContainerWidget({ ...widgetProps, steps: [step] })]
    });

    const submitButtonText = screen.getByText(widgetProps.submitButtonText);
    const multiStepsFooterContainer = document.querySelector(".form-buttons");
    const allSteps = document.querySelectorAll("div[data-step-number]");
    expect(submitButtonText).toBeInTheDocument();
    expect(multiStepsFooterContainer).not.toBeInTheDocument();
    expect(allSteps).toHaveLength(1);
  });

  it("renders with multiple steps mode", () => {
    lodashEsModule.isEmpty.mockReturnValue(false);
    const items = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    const firstStep = new FormStepWidget({ isSingleStep: false, items: items });
    const secondStep = new FormStepWidget({
      isSingleStep: false,
      items: items
    });
    pageRenderer.render({
      body: [
        new FormStepContainerWidget({
          ...widgetProps,
          steps: [firstStep, secondStep],
          formType: "multi-step"
        })
      ]
    });

    const multiStepsFooterContainer = document.querySelector(".form-buttons");
    const allSteps = document.querySelectorAll("div[data-step-number]");
    expect(multiStepsFooterContainer).toBeInTheDocument();
    expect(allSteps).toHaveLength(2);
  });

  it("shows submit & review buttons in the last step", () => {
    lodashEsModule.isEmpty.mockReturnValue(false);
    const items = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    const firstStep = new FormStepWidget({ isSingleStep: false, items: items });
    pageRenderer.render({
      body: [
        new FormStepContainerWidget({
          ...widgetProps,
          steps: [firstStep],
          formType: "multi-step"
        })
      ]
    });

    const multiStepsFooterContainer = document.querySelector(".form-buttons");
    const reviewButton = document.querySelector(".review-button");
    const submitButton = document.querySelector(".forward-button");
    const allSteps = document.querySelectorAll("div[data-step-number]");
    expect(multiStepsFooterContainer).toBeInTheDocument();
    expect(allSteps).toHaveLength(1);
    expect(reviewButton).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toHaveTextContent(widgetProps.submitButtonText);
  });

  it("shows border", () => {
    lodashEsModule.isEmpty.mockReturnValue(false);
    pageRenderer.render({
      body: [new FormStepContainerWidget({ ...widgetProps, showBorder: true })]
    });

    const container = document.querySelector(
      ".scrivito-neoletter-form-widgets.form-container-widget"
    );
    expect(container).toHaveClass("form-border");
  });

  it("renders winnie-the-pooh", () => {
    lodashEsModule.isEmpty.mockReturnValue(false);

    pageRenderer.render({
      body: [new FormStepContainerWidget(widgetProps)]
    });
    const input = document.querySelector('input[name="fax"]');
    expect(input).toBeInTheDocument();
  });

  it("renders correctly", () => {
    const items = [
      new FormDateWidget({
        customFieldName: "custom_date",
        title: "date",
        dateType: "date"
      })
    ];
    const firstStep = new FormStepWidget({ isSingleStep: false, items: items });
    const secondStep = new FormStepWidget({
      isSingleStep: false,
      items: items
    });

    const tree = pageRenderer.getAsJSON({
      body: [
        new FormStepContainerWidget({
          ...widgetProps,
          steps: [firstStep, secondStep],
          formType: "multi-step"
        })
      ]
    });
    expect(tree).toMatchSnapshot();
  });
});
