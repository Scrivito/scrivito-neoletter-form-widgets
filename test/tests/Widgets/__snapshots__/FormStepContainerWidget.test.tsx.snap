// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormStepContainerWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="scrivito-neoletter-form-widgets form-container-widget "
    >
      <form
        id="test-id"
        method="post"
      >
        <input
          name="form_id"
          type="hidden"
          value="test-id"
        />
        <input
          name="url"
          type="hidden"
        />
        <div />
        <div
          aria-hidden="true"
          className="winnie-the-pooh"
        >
          <input
            autoComplete="off"
            name="fax"
            role="presentation"
            tabIndex={-1}
          />
        </div>
        <div>
          <div>
            <div
              className=" "
              data-step-number={1}
            >
              <div
                className="row"
              >
                <div>
                  <div>
                    <div
                      className="form-date mb-3"
                    >
                      <div
                        className="date-title"
                      >
                        <span
                          className="text-super"
                        >
                          date
                        </span>
                      </div>
                      <input
                        className="datepicker"
                        onChange={[Function]}
                        required={false}
                        type="date"
                      />
                      <input
                        className="show-in-review"
                        name="custom_date"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              className=" "
              data-step-number={1}
            >
              <div
                className="row"
              >
                <div>
                  <div>
                    <div
                      className="form-date mb-3"
                    >
                      <div
                        className="date-title"
                      >
                        <span
                          className="text-super"
                        >
                          date
                        </span>
                      </div>
                      <input
                        className="datepicker"
                        onChange={[Function]}
                        required={false}
                        type="date"
                      />
                      <input
                        className="show-in-review"
                        name="custom_date"
                        type="hidden"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
      <div
        className="form-buttons"
      >
        <button
          className="btn btn-primary backward-button"
          hidden={true}
          onClick={[Function]}
        >
          back
        </button>
        <div
          className="step-counter"
        >
          1 / 2
        </div>
        <button
          className="btn btn-primary forward-button "
          hidden={false}
          onClick={[Function]}
        >
          forward
        </button>
        <button
          className="btn btn-primary submit-button"
          disabled={false}
          hidden={true}
          onClick={[Function]}
        >
          submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
